web/web.go

```diff
@@ -526,38 +531,216 @@ func dropForeignKey(w http.ResponseWriter, r *http.Request) {
+			index.Name = newName
+		}
+		newIndexes = append(newIndexes, index)
+	}
+	sp.Indexes = newIndexes
+
+	sessionState.conv.SpSchema[table] = sp
+	updateSessionFile()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(sessionState.conv)
+}
+
+//addIndexes checks the new names for spanner name validity, ensures the new names are already not used by existing tables
+//secondary indexes or foreign key constraints. If above checks passed then new indexes are added to the schema else appropriate
+//error thrown.
+func addIndexes(w http.ResponseWriter, r *http.Request) {
+	table := r.FormValue("table")
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+	}
+
+	newIndexes := []ddl.CreateIndex{}
+	if err = json.Unmarshal(reqBody, &newIndexes); err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	//Check new name for spanner name validity.
+
+	newNames := []string{}
+	newNamesMap := map[string]bool{}
+	for _, value := range newIndexes {
+		newNames = append(newNames, value.Name)
+		newNamesMap[strings.ToLower(value.Name)] = true
+	}
+	if len(newNames) != len(newNamesMap) {
+		http.Error(w, fmt.Sprintf("Conflicts exists in given new names : %s", strings.Join(newNames, ",")), http.StatusBadRequest)
+		return
+	}
+	if ok, invalidNames := checkSpannerNamesValidity(newNames); !ok {
+		http.Error(w, fmt.Sprintf("Following are not valid Spanner identifiers: %s", strings.Join(invalidNames, ",")), http.StatusBadRequest)
+		return
+	}
+
+	// Check that the new names are not already used by existing tables, secondary indexes or foreign key constraints.
+	if ok, err := canRename(newNames, table); !ok {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	sp := sessionState.conv.SpSchema[table]
+	sp.Indexes = append(sp.Indexes, newIndexes...)
+
 	sessionState.conv.SpSchema[table] = sp
 	updateSessionFile()
 	w.WriteHeader(http.StatusOK)
 	json.NewEncoder(w).Encode(sessionState.conv)
 }
 
+func checkSpannerNamesValidity(input []string) (bool, []string) {
+	status := true
+	invaliNewNames := []string{}
+	for _, newName := range input {
+		if _, status := internal.FixName(newName); status {
+			status = false
+			invaliNewNames = append(invaliNewNames, newName)
+		}
+	}
+	return status, invaliNewNames
+}
+
+func canRename(names []string, table string) (bool, error) {
+	namesMap := map[string]bool{}
+	// Check that this name isn't already used by another table.
+	for _, name := range names {
+		namesMap[name] = true
+		if _, ok := sessionState.conv.SpSchema[name]; ok {
+			return false, fmt.Errorf("new name : '%s' is used by another table", name)
+		}
+	}
+
+	// Check for this name isn't already used by another foreign key.
+	for _, sp := range sessionState.conv.SpSchema {
+		for _, foreignKey := range sp.Fks {
+			if sp.Name != table {
```

Comment:
Major:

You mentioned in an earlier comment that this is required for name swap. 

I think this is problematic since canRename is also used in addIndexes. This will let users add multiple foreign keys and multiple indexes to the same table. To simplify the implementation, I think it's better to not allow swapping the names of the constraints in the same table. (Note that we are not allowing swapping the names across multiple tables).

Reply:
Agreed, changed accordingly pl